File: 1-yahtzee-java/src/test/java/yahtzee/DiceRollerTest.java
package yahtzee;

import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.*;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.when;

public class DiceRollerTest {

    @Test
    public void testRollAll() throws Exception {

        DieRoller dieRoller = generateADieRollerWithRollingValues(1, 5, 1, 2, 1);

        DiceRoller diceRoller = new DiceRoller(dieRoller);

        diceRoller.roll(Dice.values());

        assertEquals(generateExpectedRolledDice(), diceRoller.lastRolledDice());
    }

    @Test
    public void testRollsOnlyTheDesiredDiceKeepingThePreviousValueForTheOthers() throws Exception {

        DieRoller dieRoller = generateADieRollerWithRollingValues(6, 5, 4, 2, 3, 1, 1, 1);

        DiceRoller diceRoller = new DiceRoller(dieRoller);

        diceRoller.roll(Dice.values());

        diceRoller.roll(Dice.D1, Dice.D3, Dice.D5);

        assertEquals(generateExpectedRolledDice(), diceRoller.lastRolledDice());

    }

    private DieRoller generateADieRollerWithRollingValues(Integer value, Integer... values) {
        DieRoller dieRoller = mock(DieRoller.class);
        when(dieRoller.roll()).thenReturn(value, values);
        return dieRoller;
    }

    private Map<Dice, Integer> generateExpectedRolledDice() {
        Map<Dice, Integer> rolledDice = new HashMap<Dice, Integer>();
        rolledDice.put(Dice.D1, 1);
        rolledDice.put(Dice.D2, 5);
        rolledDice.put(Dice.D3, 1);
        rolledDice.put(Dice.D4, 2);
        rolledDice.put(Dice.D5 ,1);
        return rolledDice;
    }

}

File: 2-yahtzee-php/tests/Yahtzee/DiceRollerTest.php
<?php


namespace Yahtzee;


class DiceRollerTest extends \PHPUnit_Framework_TestCase
{

    public function testRollAll()
    {
        $dieRoller = new FakeDieRoller([1, 3, 6, 2, 5]);
        $diceRoller = new DiceRoller($dieRoller);

        $diceRoller->rollAll();

        $expectedRollResult = [
            (string) Dice::D1() => 1,
            (string) Dice::D2() => 3,
            (string) Dice::D3() => 6,
            (string) Dice::D4() => 2,
            (string) Dice::D5() => 5
        ];
        $this->assertEquals($expectedRollResult, $diceRoller->lastRollResult());
    }

    public function testRerun() {
        $dieRoller = new FakeDieRoller([1, 3, 6, 2, 5,
                                        5, 4]);
        $diceRoller = new DiceRoller($dieRoller);

        $diceRoller->rollAll();
        $diceRoller->reRun([Dice::D2(), Dice::D4()]);

        $expectedRollResult = [
            (string) Dice::D1() => 1,
            (string) Dice::D2() => 5,
            (string) Dice::D3() => 6,
            (string) Dice::D4() => 4,
            (string) Dice::D5() => 5
        ];
        $this->assertEquals($expectedRollResult, $diceRoller->lastRollResult());
    }
}
